### ISSUE TITLE
Pedestrians appear unable to leave sidewalks, rendering pedestrian routing in cities poorly supported

### ISSUE THREAD

**Issue by jorisgeer** [valhalla/valhalla#4419]
In OpenStreetMap data, sidewalks are typically drawn separate from the road, even if there is no kerb in between. Thus, when routing, there is no physical connection to the road and walkers are routed all the way around any physical connection.
The concept of 'crossing a road', leaving a sidewalk, does not exist. That makes pedestrian routing in a city with many sidewalks poorly supported and comes with odd detours.
Google Maps appears not to draw sidewals as such, avoiding the issue. But once they are properly drawn, then routing software should understand how walksers walk on them and how road crossings work.

See screenshot for a typical example. The footpath is adjacent to the local road, as usual. If there are no sideways at the same side of the road, the detour can be infinite. Just as long as it takes to have that physical (in OSM terms) connection.

![Screenshot 2023-11-29 at 9 48 11 am](https://github.com/valhalla/valhalla/assets/7885507/5cf211da-bc22-46ff-bf57-4aeedf0ba080)

**Comment by nilsnolde** [comment]
Hm this is (in routing terms) a mapping error. The mapper has to make that connection. How else would we know there is one? 

It’s great OSM finally does this more or less consistently (a few years back most wouldn’t), it’s the only way to make wheelchair or blind routing really work. But IMO it needs to be done all the way and this example is missing the proper connections in OSM.

**Comment by nilsnolde** [comment]
Basically, I don't think we can ever have "arbitrary" road crossings, that's not how routing works. I know that's what one would ideally like to have, it's just hardly possible. (if both sidewalks on each road were somehow related, with a lot more work and lot less accuracy, we _could_ detect what you're drawing there, but I'd personally not be in favor; rather make sure you draw the connection in OSM, even it means more ways and is pretty much "tagging for the router", similar to synthetic ways on big squares (at least there the problem is much easier for us))

**Comment by kevinkreiser** [comment]
nils is right. in summary:

routing needs topology. there are almost infinite ways for us to programmatically make connections where they don't exist and that would make equally bad or worse routing. when the human eye detects this problem what its detecting is a problem with the data not the router.

**Comment by kevinkreiser** [comment]
ohbalso if you set a larger radius for your locations it will probably take the roads instead of the sidewalks to avoid the detour

**Comment by nilsnolde** [comment]
good point. I wonder if we should maybe set a default radius of 5 m or so in the web apps?

**Comment by kevinkreiser** [comment]
depends on the average use case. whatever we do we can't please everyone. a 5-10m radius is probably better than nothing in most cases but we are sure to hear about it when it wont snap to the road right under the pin.

**Comment by nilsnolde** [comment]
Sure, but as you say (and that was my thought too), it's the least worse default for most cases. I'll quickly PR to them.

**Comment by nilsnolde** [comment]
ref
https://github.com/openstreetmap/openstreetmap-website/pull/3848
https://github.com/gis-ops/valhalla-app/issues/204

**Comment by tomhughes** [comment]
By the way it makes life a lot easier for maintainers if you provide a link to the problem route so we can reproduce it, or at least provide the coordinates as text and not just as a picture so we can't copy and paste :-(

**Comment by tomhughes** [comment]
> Do you observe a mind shift lately @tomhughes ? It IS the only way to make routing really accessible for people with disabilities, so would be great to foster that in the community somehow (note, I'm not much in OSM forums and such)

Well in general over time people tend to map it more detail but it will hugely depend where you look and I can't see I often see it in the UK at the moment.

**Comment by tomhughes** [comment]
I believe this is the route in question:

https://www.openstreetmap.org/directions?engine=fossgis_valhalla_foot&route=-27.50773%2C152.93414%3B-27.50762%2C152.93379

**Comment by nilsnolde** [comment]
So it doesn't seem to work with increasing the radius: https://github.com/openstreetmap/openstreetmap-website/pull/4378#issuecomment-1831839944. Will look into it soon-ish, was supposed to be a quick fix;)

**Comment by kevinkreiser** [comment]
ah crap wrong issue, the other option is that sidewalks are heavily favored by default over carriageways shared with motorvehicles but i havent checked it... well looking at it doesnt seem to be any preference for sidewalk by default... need to look at other tagging on those ways to make a guess why costing would do this. loki should find that node at the intersection as a candidate if the radius is large enough. maybe call locate first just to be sure it is

**Comment by kevinkreiser** [comment]
2 options:

1. figure out why the radius doesnt find the regular automobile road
2. disable sidewalks in the import

neither option is amazing but both would provide better routing when poorly digitized sidewalk data is present in your area of interest

**Comment by jorisgeer** [comment]
> 2 options:
> 
> 1. figure out why the radius doesnt find the regular automobile road
> 2. disable sidewalks in the import
> 
> neither option is amazing but both would provide better routing when poorly digitized sidewalk data is present in your area of interest

The automobile road should often be tagged as no access for pedestrians as that is the reality. Both otions will result in either having to ignore 'foot=no' with dire consequences or having no route at all.

**Comment by kevinkreiser** [comment]
> The automobile road should often be tagged as no access for pedestrians as that is the reality.

I don't think I agree with that. in most western countries (what I'm familiar with) most roads are untagged regarding access so we fall back to computing the access based on the country its in and its highway tag. motorways in most countries should have pedestrian access disabled but motorways are a tiny minority of the edges in a graph. 

it might be interesting to try an experiment where we connect any graph nodes near a sidewalk edge that isn't already connected to the node. if the node already has other pedestrian edges. we could do it in the same way we connect transit nodes to the regular graph. the problem is there will be some cases where the connection is not physically possible in real life and we won't be able to know that. like @nilsnolde said a mapping error

**Comment by andrewharvey** [comment]
Interesting thread. I agree that pre-processing OSM data to insert additional topology connections is likely the best way to handle this. We can't expect OSM to map infinite possible places to cross the road, but we can expect OSM to map where you can't physically cross the road.

The issue is basically a desire to have a routing profile that is less strict about crossing only at designated crossings and supports crossing a road even without a designated crossing location.

For example at the location given by @jorisgeer if I want to route from A to B, a profile which only ever crosses at designated crossing locations would follow the path it currently does, but if a profile is used that allows crossing a road at an arbitary location, the optimal route would be like this:

![image](https://github.com/user-attachments/assets/10486020-d3fe-4748-9a69-e97571f38c45)

By definition of the `footway=sidewalk` tag, the way must follow alongside a `highway=*`. We could segment `footway=sidewalk` at X meter intervals and create a connection between that each segment point and the nearest `highway=*` way, only if that way does not cross another barrier way. I realise the X meter segments is an approximation, but should be good enough for most purposes. You could limit it to only doing this near an intersection which would work for this scenario, but wouldn't work when you want to route from one side of the road to the other without an intersection nearby.

You could further localise this based on local rules, for example in Australia the rule might be you can't cross the road within X meters of a dedicated crossing and must use the crossing, you can then only cross the road at a non-crossing if you're further than X meters.

I think the point around routing along the roadway vs the pedestrian way (`footway=sidewalk`) is a separate discussion, and some cases you might opt to ignore `footway=sidewalk` and only consider routing along the road network (taking into account `sidewalk` tags on the road ways, but in other cases you want more detailed routing to follow the `footway=sidewalk` in this case increasing the radius to choose to route along the road even if there is a sidewalk is not ideal.
